mqtt/mqtt-broker/src/auth/authentication_edgehub.rs

```diff
@@ -0,0 +1,338 @@
+use async_trait::async_trait;
+use serde::{Deserialize, Serialize};
+use serde_repr::Deserialize_repr;
+
+use crate::auth::{AuthId, Authenticator, Credentials};
+use crate::{AuthenticationError, Error};
+
+const API_VERSION: &str = "2020-04-20";
+
+pub struct EdgeHubAuthenticator {
+    url: String,
+}
+
+#[async_trait]
+impl Authenticator for EdgeHubAuthenticator {
+    type Error = Error;
+
+    async fn authenticate(
+        &self,
+        username: Option<String>,
+        credentials: Credentials,
+    ) -> Result<Option<AuthId>, Self::Error> {
+        self.authenticate_client((username, credentials).into())
+            .await
+            .map_err(Error::Authenticate)?
```

Comment:
Error impls From<AuthenticationError>. You shouldn't need to map this, ? will call .into() for you

Reply:
Im not sure I understand:
- the map_err() is there so it wrap the incoming AuthenticationError (a separate enum to list authentication related errors) int an Error::Authenticate(e)
- the into() is there because "authenticate_client" returns an EdgeHubAuthResponse structure, which is the mirror of the json structure coming back in the http response, and it gets turned (by the into() ) into an Option<AuthId> which is the return type of the authenticators.

Reply:
The `Authenticate(#[from] AuthenticationError)` will automatically make 
```
impl From<AuthenticationError> for Error {
    fn from(e: AuthenticationError) -> Error {
        Error::Authenticate(e)
}
```
This means you can just do 
```
self.authenticate_client((username, credentials).into())
            .await?
            .into()
```

Reply:
that worked, cool!